10539 server.rst (#10791)
bokeh/bokeh: sphinx/source/docs/user_guide/server.rst

.. _userguide_server:

Running a Bokeh server
======================

.. _userguide_server_purpose:

Purpose
-------

.. note::
    To make this guide easier to follow, consider familiarizing
    yourself with some of the core concepts of Bokeh in the section
    :ref:`userguide_concepts`.

Bokeh server makes it easy to create interactive web applications that connect
front-end UI events to running Python code.

Bokeh creates high-level Python models, such as plots, ranges, axes, and
glyphs, and then converts these objects to JSON to pass them to its client
library, BokehJS. For more information on the latter, see
:ref:`devguide_bokehjs`.

This flexible and decoupled design offers some advantages. For instance, it is

> easy to have other languages, such as R, Scala, or Lua, drive Bokeh plots and

Reviewer: Let's nix Lua from this list. The others are in various disrepair but Lua never really happened to begin with.

visualizations in the browser.

However, keeping these models in sync between the Python environment and the
browser would provide further powerful capabilities:

* respond to UI and tool events in the browser with computations or queries
  using the full power of Python
* automatically push server-side updates to the UI elements such as widgets or
  plots in the browser
* use periodic, timeout, and asynchronous callbacks to drive streaming updates

> **This is where the Bokeh server comes into play.**

Reviewer: I'm really attached the the original sentence here. I want to be 100% explicit about exactly what the driving purpose of the Bokeh server is, and that is automatic Python <---> JS synchronization. 

cc @tcmetzger for any other suggestions on how to improve things but keep this clear statement.

Reviewer 2: I'd just add a sentence after "This is where the...":

"The most important purpose of Bokeh server is synchronizing data between Python and BokehJS running in the browser."

And then continue: "Here's a simple example..."

Reviewer: OK, but my preference would be for the "The most important..." sentence to be set off and bolded by itself.

  plots in the browser
* use periodic, timeout, and asynchronous callbacks to drive streaming updates

**This is where the Bokeh server comes into play. Its primary purpose is to
synchronize data between the underlying Python environment and the BokehJS

> library running in the browser.**

Reviewer: Suggest this:
```
This is where the Bokeh server comes into play:

**The primary purpose of the Bokeh server is to synchronize data
between the underlying Python environment and the BokehJS 
library running in the browser.**
```

It probably seems like I am harping on this, but all my experience since 2016 points me towards having one clear sentence that stands alone.

    ></iframe>
    </div>

Manipulating the UI controls communicates new values to the backend via Bokeh

> server triggering callbacks that update the visuals with the input in real

Reviewer 2: This sentence is quite long, I'd recommend breaking it up. For example:

```suggestion
server. This also triggers callbacks that update the plot in the browser with the input in real
```

time.

.. _userguide_server_use_case:

Use case scenarios
------------------

Consider a few different scenarios when you might want to use the Bokeh server.

.. _userguide_server_use_case_individual:

Local or individual use
~~~~~~~~~~~~~~~~~~~~~~~

You might want to use the Bokeh server for exploratory data analysis, possibly
in a Jupyter notebook, or for a small app that you and your colleagues can run

> locally. The Bokeh server would be very convenient here, allowing for effective

Reviewer: maybe "is" instead of "would be" ?

use of the following methods:

> * :ref:`userguide_server_bokeh_client`

Reviewer: I'm not sure the server_bokeh_client ref is best here, today `bokeh.client` is for a very specific use-case of customizing individual app sessions (e.g. on a per-user basis) People used to use `bokeh.client` to try and have apps in "outside" python processes but that practice is actively anti-encouraged, and in fact some of the user-facing APIs people used for that are/have been removed. tldr we should not point people anywhere near `bokeh.client` for 'getting started"

or an internal company network. The Bokeh server also suits this usage well,
but you might want to first consult the following:

* For information on how to create Bokeh applications, see
  :ref:`userguide_server_applications`.
* For information on how to deploy a server with your application, see
  :ref:`userguide_server_deployment`.

.. _userguide_server_use_case_shared:

Shared publishing
~~~~~~~~~~~~~~~~~

Both of the scenarios above involve *one person* making applications on the
server, either for personal use or for consumption by a larger audience.

While it is possible for *several people* to publish different applications
to the same server, **this does not make for a good use case** because hosted
applications can execute arbitrary Python code. This raises process isolation
and security concerns and makes this kind of shared tenancy prohibitive.

One way to support this kind of multi-application environment with multiple
users is to build up infrastructure that can run a Bokeh server for each app or
at least for each user. The Bokeh project or a third party might create a
public service for this kind of usage in the future but such developments are

> beyond the scope this article.

Reviewer: Reads weird to me, I don't really think of parts of a projects documentation as an "article". Maybe "the project documentation" or just the original "this User's Guide"

Reviewer 2: An easy solution would be to just say "this document". 
You could also use "this documentation", I think. In my mind, the user guide has sections, so another option would be "section of the user guide".

       +---templates
            +---index.html

> Here are this tree's optional components:

Reviewer: I'm not sure everyone will immediately know what "tree" is in this context , also the optional components are really optional parts of the *application*, which just happen to be expressed by files in the directory (or not)

    N = 200

.. warning::
  The request object makes inspecting values, such as ``arguments``, easy.
  However, calling any of the Tornado methods, such as ``finish()``,  or

>   writing directly to ``request.connection`` results in undefined behavior.

Reviewer: I'd like to keep the word "unsupported" in there somewhere to make the discouragement as strong and explicit as possible.

Python callback into JavaScript. ``CustomJS`` callbacks only execute inside the
browser's JavaScript interpreter, which means that they can only interact with
JavaScript data and functions, such as BokehJS models.

Python callbacks with Jupyter interactors
'''''''''''''''''''''''''''''''''''''''''

When working with Jupyter notebooks, you can use Jupyter interactors to quickly
create simple GUI forms. Updates to GUI widgets trigger Python callbacks that
execute in the Python kernel of Jupyter. It is often useful to have these
callbacks call :func:`~bokeh.io.push_notebook` to push updates to displayed
plots. For more information, see
:ref:`userguide_jupyter_notebook_jupyter_interactors`.

.. note::
    You can push plot updates from Python to BokehJS with
    :func:`~bokeh.io.push_notebook`. For two-way communication, embed a Bokeh
    server in the notebook. For example, this lets range and selection updates
    trigger Python callbacks. For further details, see
    :bokeh-tree:`examples/howto/server_embed/notebook_embed.ipynb`

Updating from threads
'''''''''''''''''''''

You can make blocking computations in separate threads. However, you **have

> to** schedule document updates via a next tick callback. This callback executes

Reviewer: I think "must" > "have to" here (in general it's probably best to avoid English phrasal verbs as much as possible, though I had not thought about that explicitly until just now cc @tcmetzger )

Reviewer 2: I would generally try to avoid phrasal verbs because they can make reading more difficult for non-native speakers. As an alternative to "must", you could also use "required to", for example.

``bokeh serve`` on it.

.. _userguide_server_applications_hooks:

Lifecycle hooks
~~~~~~~~~~~~~~~

You may want to execute code at specific points of server or session runtime.
For instance, if you are using a Bokeh server with a Django server, you need to
call ``django.setup()`` for each Bokeh server to properly initialize Django for
use by Bokeh application code.

Bokeh enables this through a set of *lifecycle hooks*. To use these hooks,
create your application in
:ref:`directory format<userguide_server_applications_directory>` and include a
designated file called ``app_hooks.py`` in the directory. In this file you can
include any or all of the following conventionally named functions:

.. code-block:: python

    def on_server_loaded(server_context):
        '''If present, this function executes when the server starts.'''
        pass

    def on_server_unloaded(server_context):
        ''' If present, this function executes when the server shuts down. '''
        pass

    def on_session_created(session_context):
        ''' If present, this function executes when you create a session. '''
        pass

    def on_session_destroyed(session_context):

>         ''' If present, this function executes when you close a session. '''

Reviewer: I don't think "you" is appropriate in this context.  Is this "you" the app author? Or the app viewer? Ultimately, it is the Bokeh server that creates and closes sessions, and I think it would be most unambiguous more like "when the server closes a session" (and creates, above)

.. code-block:: python

    def on_server_loaded(server_context):
        '''
        If present, this function executes when the server starts.

>         '''

Reviewer: FWIW we don't have docstrings formatted like this anywhere in the codebase, we always start with a comment on the same line as the `'''`

Author: I was hesitant about changing these this way, but breaking the latter two comments mid-sentence to respect the 80-char restriction felt 'ugly'. Especially compared to how neat they looked originally. That said, I can change them to be more in line with the rest of the docs.

Author: Turns out switching to single line just about does it in terms of character limit and is more conistent with other comments in the docs.

   # start timers and services and immediately return
   server.start()

You can also create and control an ``IOLoop`` directly. This can be useful when
creating standalone "normal" Python scripts that serve Bokeh apps or embedding
a Bokeh application in a framework like Flask or Django without having to run a
separate Bokeh server process. You can find some examples of this technique in
the examples directory:

* :bokeh-tree:`examples/howto/server_embed/flask_embed.py`
* :bokeh-tree:`examples/howto/server_embed/notebook_embed.ipynb`
* :bokeh-tree:`examples/howto/server_embed/standalone_embed.py`
* :bokeh-tree:`examples/howto/server_embed/tornado_embed.py`

Also note that every command line argument for ``bokeh serve`` has a
corresponding keyword argument for ``Server``. For instance, using the

> `--allow-websocket-origin` command line argument is equivalent to passing

Reviewer 2: ```suggestion
``--allow-websocket-origin`` command line argument is equivalent to passing
```

        app.run(port=8080)

.. warning::
    You can use ``bokeh.client`` to build apps "from scratch", outside the
    Bokeh server, including running and servicing callbacks by making a
    blocking call to ``session.loop_until_closed`` in the external Python
    process. However, this usage has a number of inherent technical

>     disadvantages and is not supported.

Reviewer: `session.loop_until_closed` has already been removed in fact, I think this entire warning can be deleted.

.. _userguide_server_deployment:

Deployment scenarios
--------------------

To make your application into a user-friendly service, you'll have to deploy
your work. This subsection explores various aspects of deployment.

.. _userguide_server_deployment_standalone:

Standalone Bokeh server
~~~~~~~~~~~~~~~~~~~~~~~

You can have the Bokeh server running on a network for users to interact with
your app directly. This can be a simple solution for local network deployment,
provided the capabilities of the hardware running the server match your app
requirements and the expected number of users.

However, if you have authentication, scaling, or uptime requirements, you'll
have to consider more sophisticated deployment configurations.

> SSH Tunnels

Reviewer 2: ```suggestion
SSH tunnels
```

hostname or IP address of the gateway. The gateway may prompt you for login
credentials.

To setup the tunnel between the local machine and the gateway, run the
following command on the **local machine**:

.. code-block:: sh

    ssh -NfL localhost:5006:localhost:5006 user@gateway.host

Again, replace *user* with your username on the gateway and *gateway.host*
with the hostname or IP address of the gateway.

You should now be able to access the Bokeh server from the local machine by
navigating to ``localhost:5006``. You can even set up client connections from
a Jupyter notebook running on the local machine.

.. note::
    We intend to expand this section with more guidance for other tools and
    configurations. If you have experience with other web deployment scenarios
    and wish to contribute your knowledge here, please contact us on
    https://discourse.bokeh.org

.. _userguide_server_deployment_ssl:

SSL termination
~~~~~~~~~~~~~~~

You can configure the Bokeh server to terminate SSL connections and serve
secure HTTPS and WSS sessions directly. To do so, you'll have to supply the
``--ssl-certfile`` argument with the value of the path to a single PEM file

> containing a certificate as well as any number of CA certificates needed to

Reviewer: Might be nice for "CA certificates" to link to https://en.wikipedia.org/wiki/Certificate_authority or similar

    }

This configuration will proxy all incoming HTTPS connections to ``foo.com``
over to a Bokeh server running internally on ``http://127.0.0.1:5100``.

.. _userguide_server_deployment_nginx_load_balance:

Load balancing with Nginx
~~~~~~~~~~~~~~~~~~~~~~~~~

The Bokeh server is scalable by design. If you need more capacity, you can
simply run additional servers. In this case, you'll generally want to run all
the Bokeh server instances behind a load balancer so that new connections are
distributed among individual servers.

.. figure:: /_images/bokeh_serve_scale.svg
    :align: center
    :width: 65%

    The Bokeh server is horizontally scalable. To add more capacity, you
    can run more servers behind a load balancer.

Nginx can help with load balancing. This section describes some of the basics
of one possible configuration, but please also refer to the
`Nginx load balancer documentation`_. For instance, there are different
strategies available for choosing what server to connect to next.

> First, we need to add an ``upstream`` stanza to our Nginx configuration.

Reviewer 2: ```suggestion
First, add an ``upstream`` stanza to your Nginx configuration.
```

        pass

If the module provides a relative ``login_url``, it can also provide an
optional ``LoginHandler`` class, which the Bokeh server will incorporate
automatically.

The ``get_login_url`` function is useful in cases where the login URL must
vary based on the request, cookies, or other factors. You can also specify a
``LoginHandler`` when defining the ``get_url_function``.

To define an endpoint for logging users out, you can also use optional
``logout_url`` and ``LogoutHandler`` parameters, similar to the login options.

If you don't provide an authentication module, the configuration will not
require any authentication to access Bokeh server endpoints.

.. warning::
    The configuration executes the contents of the authentication module.

Secure cookies
''''''''''''''

If you want to use Tornado's `set_secure_cookie`_ and `get_secure_cookie`_
functions in your auth module, you'll have to set a cookie secret. To do so,
use the ``BOKEH_COOKIE_SECRET`` environment variable.

.. code-block:: sh

    export BOKEH_COOKIE_SECRET=<cookie secret value>

The value should be a long, random sequence of bytes.

.. _userguide_server_deployment_security:

Security
~~~~~~~~

> By default, the Bokeh server will accept any incoming connections with an

Reviewer: cc @tcmetzger for thoughts

I had used "a Bokeh server" here to emphasize "a concrete, actually running Bokeh server process" as opposed to what (to me) reads more like an abstract concept of "the Bokeh server" but maybe I am being pedantic.

Reviewer 2: I'd be ok with "the Bokeh Server". We are talking about the server's default behavior, and allowing those connections seems to be true for any instance of a Bokeh server (in its default configuration) - this could therefore be considered a property of the abstract concept of "the Bokeh server".

Both of these open your default browser to the default application URL
``localhost:5006`` and, since ``localhost:5006`` is on the list of allowed
WebSocket origins, the Bokeh server creates and displays a new session.

When you embed a Bokeh server in another web page with |server_document| or
|server_session|, the ``Origin`` header for the request to the Bokeh server
is the URL of the page that hosts your Bokeh content.

For example, if a user navigates to your page at ``https://acme.com/products``,
the origin header reported by the browser will be ``acme.com``. In this case,
you'd typically restrict the Bokeh server to honoring *only* the requests that
originate from the ``acme.com`` page, preventing other pages from embedding
your Bokeh app without your knowledge.

You can do so by setting the ``--allow-websocket-origin`` command line argument
as follows:

.. code-block:: sh

    bokeh serve --show --allow-websocket-origin=acme:com myapp.py

This will prevent other sites from embedding your Bokeh application in their
pages because requests from users viewing those pages will report a different
origin than ``acme.com``, causing the Bokeh server to reject them.

.. warning::

>     Bear in mind that this only prevents *other web pages* from sneaking off

Reviewer: "sneaking off" is too colloquial I think. I think the part to get across "...from embedding your Bokeh app in their site without your permission" I'm not even sure the bits about spoofing origin headers need to be kept (or that this even needs to be a warning)

    with your Bokeh app and displaying it to an audience using standard web
    browsers. A determined and knowledgeable attacker can spoof Origin headers.

If you require multiple allowed origins, you can pass multiple instances of
``--allow-websocket-origin`` on the command line.

You can also configure the Bokeh server to allow all connections regardless of
origin:

.. code-block:: sh

    bokeh serve --show --allow-websocket-origin='*' myapp.py

> This option should only be used for testing and experimentation.

Reviewer: I'd like to avoid a "should" here. Our advice is to use this setting only for testing, experimentation, or local notebook usage. (However best to express that).

                            session_id=generate_session_id())
    return render_template("embed.html", script=script, template="Flask")

Make sure to set identical ``BOKEH_SECRET_KEY`` environment variables both for

> the Bokeh server and for the web app processes, be it Flask, Django, or any

Reviewer: "be it ..." rings a bit unusual to me. Maybe just just delete the last part after "web app process" entirely?

other tool you are using.

.. note::

    Signed session IDs serve as access tokens. As with any token system,
    security is predicated on keeping the token secret. You should also run

>     the Bokeh server behind a proxy that terminates SSL connections. This lets

Reviewer: Bokeh server can now terminate its own SSL connections so maybe add ", or configure the Bokeh server to terminate SSL directly"
